### PR TITLE
Unroll w vectors

### PR DESCRIPTION
**Pull request by cgmillette**
- Replace std::tuple instances with non_native_vector_base (constexpr friendly)
- Update the hipRTC test case
- Ensure hipRTC can build and validate properly

### CODE REVIEW

File: library/include/rocwmma/internal/vector.hpp [hunk 1]
```diff
@@ -404,6 +418,228 @@ namespace rocwmma
...
+        constexpr static inline uint32_t size()
+        {
+            return VecSize;
+        }
+    };
+
+    namespace detail
+    {
+        template <typename T, typename... Ts>
+        constexpr auto getFirstType()
```
Comment (CongMa13) [inline library/include/rocwmma/internal/vector.hpp]: The name implies it returns a type. Do you think we can merge line : 545 into this function? 

`using DataT = decltype(detail::getFirstType<std::decay_t<Ts>...>());`

File: library/include/rocwmma/internal/tuple.hpp [hunk 2]
```diff
@@ -26,341 +26,283 @@
...
-    template <typename... TypesL, typename... TypesR, size_t... Indices>
-    constexpr static inline auto operator_mult_impl(tuple<TypesL...> const& lhs,
-                                                    tuple<TypesR...> const& rhs,
-                                                    index_sequence<Indices...>)
+    template <typename T, unsigned int n, typename U>
+    ROCWMMA_HOST_DEVICE inline constexpr non_native_vector_base<T, n>
+        operator-(U y, const non_native_vector_base<T, n>& x) noexcept
     {
-        return make_tuple(get<Indices>(lhs) * get<Indices>(rhs)...);
+        return non_native_vector_base<T, n>{x} -= non_native_vector_base<T, n>{y};
```
Comment (CongMa13) [inline library/include/rocwmma/internal/tuple.hpp]: Should `y - x` be 
```
return non_native_vector_base<T, n>{y} -= non_native_vector_base<T, n>{x};
```

Maybe we should keep the unit test for `operator -`

File: library/include/rocwmma/internal/tuple.hpp [hunk 3]
```diff
@@ -26,341 +26,283 @@
...
-    template <typename... Types, size_t... Indices>
-    constexpr static inline auto operator_add_impl(tuple<Types...> const& lhs,
-                                                   tuple<Types...> const& rhs,
-                                                   index_sequence<Indices...>)
+    template <typename T, unsigned int n, typename U>
+    ROCWMMA_HOST_DEVICE inline constexpr non_native_vector_base<T, n>
+        operator/(U y, const non_native_vector_base<T, n>& x) noexcept
     {
-        return make_tuple(get<Indices>(lhs) + get<Indices>(rhs)...);
+        return non_native_vector_base<T, n>{x} /= non_native_vector_base<T, n>{y};
```
Comment (CongMa13) [inline library/include/rocwmma/internal/tuple.hpp]: Should `y / x` be 
```
return non_native_vector_base<T, n>{y} /= non_native_vector_base<T, n>{x};
```

File: test/unit/tuple_test/device/tuple.hpp [hunk 4]
```diff
@@ -30,113 +30,32 @@
...
 
-        auto srcTuple          = std::make_tuple(1, 2.0, 3u);
-        auto expectHeadElement = std::make_tuple(2);
-        auto resultHeadElement = std::operator_mult_impl(2, srcTuple, std::index_sequence<0>{});
-        err |= expectHeadElement != resultHeadElement;
+        auto const srcTuple          = make_vector(1, 2, 3);
+        auto       expectHeadElement = make_vector(2);
+        auto       resultHeadElement = detail::mult_poly_vec_impl(
+            make_vector(2, 2, 2), srcTuple, detail::index_sequence<0>{});
+        err |= vector_reduce_and(expectHeadElement != resultHeadElement);
```
Comment (CongMa13) [inline test/unit/tuple_test/device/tuple.hpp:49]: Do we need to add `vector_reduce_and` around a Boolean result?

File: test/unit/tuple_test/device/tuple.hpp [hunk 5]
```diff
@@ -30,113 +30,32 @@
...
 
-        auto expectTailElement = std::make_tuple(4.0, 6u);
-        auto resultTailElement = std::operator_mult_impl(2, srcTuple, std::index_sequence<1, 2>{});
-        err |= expectTailElement != resultTailElement;
+        auto expectTailElement = make_vector(4, 6);
+        auto resultTailElement = detail::mult_poly_vec_impl(
+            make_vector(2, 2, 2), srcTuple, detail::index_sequence<1, 2>{});
+        err |= vector_reduce_and(expectTailElement != resultTailElement);
```
Comment (CongMa13) [inline test/unit/tuple_test/device/tuple.hpp:54]: Same for all `vector_reduce_and` in the test file.

File: library/include/rocwmma/internal/tuple.hpp (hunk 2 above)
Comment (cgmillette) [reply library/include/rocwmma/internal/tuple.hpp]: Nice catch, tnx!

File: library/include/rocwmma/internal/vector.hpp (hunk 1 above)
Comment (cgmillette) [reply library/include/rocwmma/internal/vector.hpp]: Moved to a more useful utility as first_type_t

File: test/unit/tuple_test/device/tuple.hpp (hunk 4 above)
Comment (cgmillette) [reply test/unit/tuple_test/device/tuple.hpp:49]: Vector operations have a boolean vector result. These vectors need to be reduced

**Review by CongMa13 — APPROVED** [review]
Awesome~

**Pull request merged**